modules/tensor_mechanics/src/materials/ComputeCappedWeakPlaneStress.C

```diff
@@ -0,0 +1,890 @@
+    _tan_psi(getUserObject<TensorMechanicsHardeningModel>("tan_dilation_angle")),
+    _tstrength(getUserObject<TensorMechanicsHardeningModel>("tensile_strength")),
+    _cstrength(getUserObject<TensorMechanicsHardeningModel>("compressive_strength")),
+    _small_smoother2(Utility::pow<2>(getParam<Real>("tip_smoother"))),
+    _max_nr_its(getParam<unsigned>("max_NR_iterations")),
+    _perform_finite_strain_rotations(getParam<bool>("perform_finite_strain_rotations")),
+    _smoothing_tol(getParam<Real>("smoothing_tol")),
+    _f_tol(getParam<Real>("yield_function_tol")),
+    _f_tol2(Utility::pow<2>(getParam<Real>("yield_function_tol"))),
+    _tangent_operator_type((TangentOperatorEnum)(int)getParam<MooseEnum>("tangent_operator")),
+    _perfect_guess(getParam<bool>("perfect_guess")),
+    _stress_return_type(nothing_special),
+    _min_step_size(getParam<Real>("min_step_size")),
+    _step_one(declareRestartableData<bool>("step_one", true)),
+
+    _plastic_strain(declareProperty<RankTwoTensor>("plastic_strain")),
+    _plastic_strain_old(declarePropertyOld<RankTwoTensor>("plastic_strain")),
+    _intnl(declareProperty<std::vector<Real> >("plastic_internal_parameter")),
+    _intnl_old(declarePropertyOld<std::vector<Real> >("plastic_internal_parameter")),
+    _yf(declareProperty<std::vector<Real> >("plastic_yield_function")),
+    _iter(declareProperty<Real>("plastic_NR_iterations")), // this is really an unsigned int, but for visualisation i convert it to Real
+    _linesearch_needed(declareProperty<Real>("plastic_linesearch_needed")), // this is really a boolean, but for visualisation i convert it to Real
+
+    _strain_increment(getMaterialPropertyByName<RankTwoTensor>(_base_name + "strain_increment")),
+    _rotation_increment(getMaterialPropertyByName<RankTwoTensor>(_base_name + "rotation_increment")),
+
+    _stress_old(declarePropertyOld<RankTwoTensor>(_base_name + "stress")),
+    _elastic_strain_old(declarePropertyOld<RankTwoTensor>(_base_name + "elastic_strain")),
+
+    _dgaE_dpt(0.0),
+    _dp_dpt(0.0),
+    _dq_dpt(0.0),
+    _dgaE_dqt(0.0),
+    _dp_dqt(0.0),
+    _dq_dqt(0.0),
+    _rhs(3),
+    _all_q(3)
+{
+  // With arbitary UserObjects, it is impossible to check everything,
+  // but this will catch the common errors
+  if (_tan_phi.value(0) < 0 || _tan_psi.value(0) < 0)
+    mooseError("ComputeCappedWeakPlaneStress: Weak-plane friction and dilation angles must lie in [0, Pi/2]");
+  if (_tan_phi.value(0) < _tan_psi.value(0))
+    mooseError("ComputeCappedWeakPlaneStress: Weak-plane friction angle must not be less than dilation angle");
+  if (_cohesion.value(0) < 0)
+    mooseError("ComputeCappedWeakPlaneStress: Weak-plane cohesion must not be negative");
+  if (_tstrength.value(0) < - _cstrength.value(0))
+    mooseError("ComputeCappedWeakPlaneStress: Weak plane tensile strength must not be less than negative-compressive-strength");
+
+  for (unsigned i = 0; i < 3; ++i)
```

Comment:
I don't like seeing bare constants like `3`. Can we use LIBMESH_DIM?

Minor: Since this vector is fixed at compile time you can consider using the new `std::array()` class. Basically this is the new C++11 wrapper around statically allocated C-arrays but with a C++ like feel. Why would you use them? They live on the stack instead of going through the memory allocator for storage on the free-store so they can potentially be a lot faster.